cql3/expr/expression.cc

```diff
@@ -1290,5 +1294,210 @@ std::vector<expression> extract_single_column_restrictions_for_column(const expr
 }
 
 
+constant::constant(cql3::raw_value val, data_type typ)
+    : value(std::move(val)), type(std::move(typ)) {
+}
+
+constant constant::make_null(data_type val_type) {
+    return constant(cql3::raw_value::make_null(), std::move(val_type));
+}
+
+constant constant::make_unset_value(data_type val_type) {
+    return constant(cql3::raw_value::make_unset_value(), std::move(val_type));
+}
+
+constant constant::make_bool(bool bool_val) {
+    return constant(raw_value::make_value(boolean_type->decompose(bool_val)), boolean_type);
+}
+
+bool constant::is_null() const {
+    return value.is_null();
+}
+
+bool constant::is_unset_value() const {
+    return value.is_unset_value();
+}
+
+bool constant::has_empty_value_bytes() const {
+    if (is_null_or_unset()) {
+        return false;
+    }
+
+    return value.to_view().size_bytes() == 0;
+}
+
+bool constant::is_null_or_unset() const {
+    return is_null() || is_unset_value();
+}
+
+std::optional<bool> get_bool_value(const constant& constant_val) {
+    if (constant_val.type->get_kind() != abstract_type::kind::boolean) {
+        return std::nullopt;
+    }
+
+    if (constant_val.is_null_or_unset()) {
+        return std::nullopt;
+    }
+
+    if (constant_val.has_empty_value_bytes()) {
+        return std::nullopt;
+    }
+
+    return constant_val.value.to_view().deserialize<bool>(*boolean_type);
+}
+
+constant evaluate(term* term_ptr, const query_options& options) {
+    if (term_ptr == nullptr) {
+        return constant::make_null();
+    }
+
+    ::shared_ptr<terminal> bound = term_ptr->bind(options);
+    if (bound.get() == nullptr) {
+        return constant::make_null();
+    }
+
+    raw_value raw_val = bound->get(options);
+    data_type val_type = bound->get_value_type();
+    return constant(std::move(raw_val), std::move(val_type));
+}
+
+constant evaluate(const ::shared_ptr<term>& term_ptr, const query_options& options) {
+    return evaluate(term_ptr.get(), options);
+}
+
+constant evaluate(term& term_ref, const query_options& options) {
+    return evaluate(&term_ref, options);
+}
+
+cql3::raw_value_view evaluate_to_raw_view(const ::shared_ptr<term>& term_ptr, const query_options& options) {
+    constant value = evaluate(term_ptr, options);
+    return cql3::raw_value_view::make_temporary(std::move(value.value));
+}
+
+cql3::raw_value_view evaluate_to_raw_view(term& term_ref, const query_options& options) {
+    constant value = evaluate(term_ref, options);
+    return cql3::raw_value_view::make_temporary(std::move(value.value));
+}
+
+// A function used to get elements from a constant value.
+// As arguments takes:
+// * val - The value to get elements from.
+// * expected_type_kind - The expected type of this value.
+// * can_have_0_bytes - Whether a 0 byte (empty) value is a valid value.
+//                      This is true for example for tuples, but a list must always have some bytes to specify size.
```

Comment:
I think so, [the spec](https://github.com/apache/cassandra/blob/trunk/doc/native_protocol_v4.spec#L960) allows it:
```
6.20 tuple

  A sequence of [bytes] values representing the items in a tuple.  The encoding
  of each element depends on the data type for that position in the tuple.
  Null values may be represented by using length -1 for the [bytes]
  representation of an element.
  
7. User Defined Types

  This section describes the serialization format for User defined types (UDT),
  as described in section 4.2.5.2.

  A UDT value is composed of successive [bytes] values, one for each field of the UDT
  value (in the order defined by the type). A UDT value will generally have one value
  for each field of the type it represents, but it is allowed to have less values than
  the type has fields.
```

We can achieve an empty tuple value by doing `WHERE pk in ()`.
I'm not sure if it's possible to easily achieve empty user type value.

Reply:
I don't agree with your reading of the spec, because of this paragraph at the beginning of section 6:

["All values are represented as [bytes] in EXECUTE and RESULT messages.
  The [bytes] format includes an int prefix denoting the length of the value.
  For that reason, the serialization formats described here will not include
  a length component."](https://github.com/apache/cassandra/blob/138569b079b3d17b1020a24463adabecd903b79f/doc/native_protocol_v4.spec#L856-L859)

Reply:
No that's the length of total bytes. This is the `raw_value_view::size_bytes()`.
Inside those bytes collections have an `int32` for size, while tuples have just the elements.

Reply:
OK.  I'd still skip this parameter and check the type inside the callee.